### PR TITLE
[CDAP-19218] Distinguish between user and system programs in master environment extensions

### PR DESCRIPTION
**Pull request by dli357**

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 1]
```diff
@@ -358,6 +366,9 @@ public TwillPreparer withConfiguration(Map<String, String> config) {
     if (config.containsKey(KubeTwillRunnerService.RUNTIME_CLEANUP_DISABLED)) {
       runtimeCleanupDisabled = Boolean.parseBoolean(config.get(KubeTwillRunnerService.RUNTIME_CLEANUP_DISABLED));
     }
+    if (config.containsKey(MasterOptionConstants.RUNTIME_NAMESPACE)) {
+      cdapRuntimeNamespace = config.get(MasterOptionConstants.RUNTIME_NAMESPACE);
```
Comment (albertshau) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java:370]: how is this different than the programRuntimeNamespace above?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 2]
```diff
@@ -1178,26 +1189,30 @@ private V1Container createContainer(String name, String containerImage, String i
...
     float memoryMultiplier = Float.parseFloat(cConf.getOrDefault(MEMORY_MULTIPLIER, DEFAULT_MULTIPLIER));
 
     V1ResourceRequirementsBuilder requirementsBuilder = new V1ResourceRequirementsBuilder();
 
+    LOG.info("Getting resource requirements for CDAP namespace {}...", cdapRuntimeNamespace);
     // Use separate multiplier for user pods
-    if (resourceType == V1Job.class) {
+    // System pods have the 'system' namespace
+    if (cdapRuntimeNamespace != null && !isSystemNamespace(cdapRuntimeNamespace)) {
+      LOG.info("Using program multiplier for non-system workloads");
```
Comment (albertshau) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: let's log the multiplier too

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/PropertiesResolver.java [hunk 3]
```diff
@@ -74,6 +74,7 @@ public Map<String, String> getSystemProperties(ProgramId id)
     throws IOException, NamespaceNotFoundException, AccessException {
     Map<String, String> systemArgs = new HashMap<>();
     systemArgs.put(Constants.CLUSTER_NAME, cConf.get(Constants.CLUSTER_NAME, ""));
+    systemArgs.put(ProgramOptionConstants.RUNTIME_NAMESPACE, id.getNamespace());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/PropertiesResolver.java]: I don't think we need to put it here, as the programId is passed all the way down as well.

Can we add it directly to the twillConfigs in the DistributedProgramRunner before calling withConfiguration()?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 2 above)
Comment (dli357) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/PropertiesResolver.java (hunk 3 above)
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/PropertiesResolver.java]: Fixed

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 1 above)
Comment (dli357) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java:370]: programRuntimeNamespace is the kubernetes namespace in which the program runs in which does not necessarily match the CDAP namespace

**Review by albertshau — APPROVED** [review]

**Pull request merged**